hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1031 @@
+    xedInstrRR(XED_ICLASS_MOVQ, rSrc, rdest);
+  }
+
+  void addsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_ADDSD, srcdest, src);
+  }
+  void mulsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_MULSD, srcdest, src);
+  }
+  void subsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_SUBSD, srcdest, src);
+  }
+  void pxor(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_PXOR, srcdest, src);
+  }
+  void cvtsi2sd(Reg64 src, RegXMM dest) {
+    xedInstrRR(XED_ICLASS_CVTSI2SD, XED_INVERSE(dest, src));
+  }
+  void cvtsi2sd(MemoryRef m, RegXMM dest) {
+    xedInstrMR(XED_ICLASS_CVTSI2SD, m, dest);
+  }
+  void ucomisd(RegXMM l, RegXMM r) {
+    xedInstrRR(XED_ICLASS_UCOMISD, l, r);
+  }
+  void sqrtsd(RegXMM src, RegXMM dest) {
+    xedInstrRR(XED_ICLASS_SQRTSD, dest, src);
+  }
+
+  void divsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_DIVSD, srcdest, src);
+  }
+  void cvttsd2siq(RegXMM src, Reg64 dest) {
+    xedInstrRR(XED_ICLASS_CVTTSD2SI, XED_INVERSE(dest, src));
+  }
+
+private:
+  //Xed conversion funcs
+#define CC_TO_XED_ARRAY(xed_instr) {                            \
+    XED_ICLASS_##xed_instr##O,    /*CC_O                  */    \
+    XED_ICLASS_##xed_instr##NO,   /*CC_NO                 */    \
+    XED_ICLASS_##xed_instr##B,    /*CC_B, CC_NAE          */    \
+    XED_ICLASS_##xed_instr##NB,   /*CC_AE, CC_NB, CC_NC   */    \
+    XED_ICLASS_##xed_instr##Z,    /*CC_E, CC_Z            */    \
+    XED_ICLASS_##xed_instr##NZ,   /*CC_NE, CC_NZ          */    \
+    XED_ICLASS_##xed_instr##BE,   /*CC_BE, CC_NA          */    \
+    XED_ICLASS_##xed_instr##NBE,  /*CC_A, CC_NBE          */    \
+    XED_ICLASS_##xed_instr##S,    /*CC_S                  */    \
+    XED_ICLASS_##xed_instr##NS,   /*CC_NS                 */    \
+    XED_ICLASS_##xed_instr##P,    /*CC_P                  */    \
+    XED_ICLASS_##xed_instr##NP,   /*CC_NP                 */    \
+    XED_ICLASS_##xed_instr##L,    /*CC_L, CC_NGE          */    \
+    XED_ICLASS_##xed_instr##NL,   /*CC_GE, CC_NL          */    \
+    XED_ICLASS_##xed_instr##LE,   /*CC_LE, CC_NG          */    \
+    XED_ICLASS_##xed_instr##NLE   /*CC_G, CC_NLE          */    \
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedJump(ConditionCode c) {
+    assert(c != CC_None);
+    static xed_iclass_enum_t jumps[] = CC_TO_XED_ARRAY(J);
+    return jumps[(int)c];
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedSetCC(int c) {
+    assert(c != -1);
+    static xed_iclass_enum_t setccs[] = CC_TO_XED_ARRAY(SET);
+    return setccs[c];
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedCMov(ConditionCode c) {
+    assert(c != CC_None);
+    static xed_iclass_enum_t cmovs[] = CC_TO_XED_ARRAY(CMOV);
+    return cmovs[(int)c];
+  }
+
+  //Xed Emit funcs
+
+#define DECLARE_UNUSED(type, name)  type name; static_cast<void>(name)
+
+#define XED_EMIT_BEGIN()                                                    \
```

Comment:
This should be doable without macros by having one `xedEmit` function that has the prologue and epilogue, and takes a lambda that does the call to the right `xed_inst*()`.